### PR TITLE
fix: Seamless AWS IAM User credential recovery with generic prompting interface

### PR DESCRIPTION
**Pull request by aknysh**
## what

- Implement generic credential prompting interface (`CredentialPromptSpec`, `CredentialField`) for multi-cloud extensibility
- Add AWS IAM User implementation that uses the generic interface
- Automatically detect `InvalidClientTokenId` error when AWS access keys are rotated or revoked
- Clear stale credentials from keyring and prompt for new ones inline during `atmos auth login`
- Fix session duration bug where configured 36h MFA sessions expired after 12h
- Add comprehensive error detection for `ExpiredTokenException` and `AccessDenied` with proper explanations and hints
- Add MFA-only re-prompt flow - when MFA token is invalid but long-lived credentials are still valid, only re-prompt for MFA token (not all credentials)
- Detect session credentials accidentally stored in keyring and prompt for new long-lived credentials
- Add warning message to `auth whoami` when credentials are invalid or expired with recovery instructions
- Add helpful tip to `auth exec` when subprocess fails, guiding users to refresh credentials
- Fix `auth whoami` to display proper session token expiration by preferring session credentials from files over long-lived keyring credentials
- Comprehensive test coverage (81.1% for `pkg/auth/identities/aws`)

## why

- Users experienced persistent authentication failures after AWS credential rotation
- `atmos auth logout` + `atmos auth login` didn't resolve the issue because stale credentials remained in keyring
- Session duration from `atmos auth user configure` was not being preserved
- When MFA token expired, users had to re-enter ALL credentials instead of just the MFA token
- Session credentials accidentally stored in keyring would cause `InvalidClientTokenId` errors on `GetSessionToken` calls
- `auth whoami` showed incorrect authentication status without guidance on how to fix it
- `auth exec` failures left users guessing about the cause when credentials were expired
- `auth whoami` was not showing "Expires" field because it loaded long-lived credentials from keyring instead of session credentials from files
- Need extensible architecture for future Azure/GCP credential prompting support

### Single Command Recovery
```shell
$ atmos auth login dev-admin

⚠ AWS credentials are required for identity: dev-admin

AWS Access Key ID: AKIAXXXXXXXXXX
AWS Secret Access Key: ********
MFA ARN (optional): arn:aws:iam::123456789012:mfa/user
Session Duration (optional, default: 12h): 36h

✓ Credentials saved to keyring: dev-admin

Enter MFA Token: 123456

✓ Authentication successful!
```

### MFA-Only Re-prompt (Session Expired)
When session expires but long-lived credentials are still valid:
```shell
$ atmos auth login dev-admin

Enter MFA Token: 123456    # Invalid/expired token

⚠ MFA token was invalid, prompting for new token

Enter MFA Token: 789012    # User enters new token

✓ Authentication successful!
```

### Improved Whoami Status
The `auth whoami` command now shows session token expiration and displays a warning with recovery instructions when credentials are invalid:

```shell
$ atmos auth whoami dev-admin
✗ Current Authentication Status

  Provider      aws-user
  Identity      dev-admin
  Expires       2025-12-30 10:11:05 EST (expired)
  Last Updated  2025-12-30 09:55:34 EST

⚠ Credentials may be expired or invalid.
  Run 'atmos auth login --identity dev-admin' to refresh.
```

### Auth Exec Guidance
When `auth exec` runs a command that fails due to expired credentials, it now provides a helpful tip:

```shell
$ atmos auth exec --identity dev-admin -- aws sts get-caller-identity
An error occurred (ExpiredToken) when calling the GetCallerIdentity operation: The security token included in the request is expired

 Tip  If credentials are expired, refresh with:
      atmos auth login --identity dev-admin
```

## Error Detection and Response

| Error Code | Meaning | Action |
|------------|---------|--------|
| `InvalidClientTokenId` | Access keys rotated/revoked | Clear stale credentials, prompt for new ones, retry |
| `ExpiredTokenException` | Session token expired | Guide user to re-login |
| `AccessDenied` (MFA-related) | Invalid/expired MFA token | Re-prompt for MFA token only, retry |
| `AccessDenied` (permission) | Missing IAM permissions | Guide user to check IAM policies |

## Architecture

Introduces a generic credential prompting interface:

```go
// pkg/auth/types/credential_prompt.go
type CredentialField struct {
    Name, Title, Description string
    Required, Secret         bool
    Default                  string
    Validator                func(string) error
}

type CredentialPromptSpec struct {
    IdentityName string
    CloudType    string  // "aws", "azure", "gcp"
    Fields       []CredentialField
}

type CredentialPromptFunc func(spec CredentialPromptSpec) (map[string]string, error)
```

Each identity type (AWS IAM User, Azure, GCP) can define its own credential fields, and the prompting UI is generic.

### Session Credential Loading

The auth manager intelligently loads credentials:
1. First checks keyring for cached credentials
2. If keyring has long-lived credentials (no session token), also checks files for session credentials
3. Prefers session credentials from files when available (they have proper expiration info)
4. Falls back to keyring credentials if no session credentials exist in files

This ensures `auth whoami` displays accurate expiration times for session tokens.

## Applies To

This enhancement applies to **AWS IAM User** identities (`aws/user` kind). Other identity types like AWS SSO, assume-role, and permission-set are not affected.

## Test Coverage

- `pkg/auth/identities/aws`: 81.1% statement coverage
- All error scenarios have unit test coverage
- Integration tests with mock auth provider
- Tests for `loadCredentialsWithFallback` session credential preference
- Tests for `loadSessionCredsFromFiles` edge cases
- Tests for `printWhoamiHuman` warning message
- Tests for `printAuthExecTip` guidance message

## Files Modified

| File | Changes |
|------|---------|
| `cmd/auth_whoami.go` | Added warning message when credentials are invalid |
| `cmd/auth_exec.go` | Added tip message when subprocess fails |
| `pkg/auth/manager.go` | Added `loadCredentialsWithFallback` and `loadSessionCredsFromFiles` for session credential preference |
| `pkg/auth/manager_extended_test.go` | Added 6 tests for session credential loading |
| `cmd/auth_whoami_test.go` | Added 2 tests for warning message |
| `cmd/auth_exec_test.go` | Added 1 test (2 subtests) for tip message |
| `docs/prd/auth-credential-invalidation-handling.md` | Updated PRD with new features |
| `website/blog/2025-12-22-auth-credential-invalidation-recovery.mdx` | Updated blog post with new features |

## Summary by CodeRabbit

* **New Features**
  * Inline credential prompting and single-command recovery for invalid/rotated AWS credentials, with MFA re-prompting and session-duration controls
  * Non-interactive contexts suppress prompts; session tokens preserved without overwriting long-lived credentials

* **Bug Fixes**
  * Better detection and user guidance for common AWS auth errors; tips shown on auth-exec failures

* **Documentation**
  * New PRD and blog post explaining recovery flows and UX

* **Tests**
  * Extensive test coverage for auth flows, STS errors, prompting, and caching behavior

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by aknysh** [comment]
@coderabbitai full review please

File: cmd/auth_credential_prompt.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package cmd
```
Comment (osterman) [inline cmd/auth_credential_prompt.go]: This belongs in the auth pkg no?

File: cmd/auth_credential_prompt.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	}, nil
+}
+
+// buildCredentialFormFields creates the form fields for credential input.
+func buildCredentialFormFields(accessKeyID, secretAccessKey, mfaArnInput, sessionDuration *string, mfaArn string) []huh.Field {
+	var formFields []huh.Field
+
+	// Access Key ID - required.
+	formFields = append(formFields, huh.NewInput().
+		Title("AWS Access Key ID").
```
Comment (osterman) [inline cmd/auth_credential_prompt.go]: This is leaking AWS details into what should be a generic interface that works with callbacks for each cloud provider.

File: docs/prd/auth-credential-invalidation-handling.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
+# Product Requirements Document: Auth Credential Invalidation Handling
+
+## Executive Summary
+
+This PRD addresses the issue where `atmos auth login` fails persistently after AWS credentials have been rotated or revoked on the AWS side. The solution automatically detects the `InvalidClientTokenId` error from AWS STS, clears stale credentials from the keyring, prompts for new credentials inline, and retries authentication.
```
Comment (osterman) [inline docs/prd/auth-credential-invalidation-handling.md]: Reframe the PRD into what we want to achieve and why, rather than framed around a bug or issue.

File: errors/errors.go [hunk 4]
```diff
@@ -556,6 +556,10 @@ var (
 	ErrSSOProvisioningFailed  = errors.New("aws sso identity provisioning failed")
 	ErrSSOInvalidToken        = errors.New("invalid aws sso token")
 
+	// AWS IAM User credential errors.
+	ErrAwsCredentialsInvalid = errors.New("aws credentials are invalid or have been revoked")
```
Comment (osterman) [inline errors/errors.go]: We can probably generalize this to be just an error for invalid credentials or revoked

File: pkg/auth/identities/aws/user.go [hunk 5]
```diff
@@ -223,76 +235,181 @@ func (i *userIdentity) writeAWSFiles(creds *types.AWSCredentials, region string)
...
+		logKeyIdentity, i.name, "error_code", apiErr.ErrorCode(), "suggestion", "reconfigure credentials")
+
+	i.clearStaleCredentials()
+
+	if PromptCredentialsFunc != nil {
+		return i.promptForNewCredentials(apiErr.ErrorCode())
+	}
+
+	return nil, errUtils.Build(errUtils.ErrAwsCredentialsInvalid).
+		WithHint("Your AWS access keys have been rotated or revoked on the AWS side").
```
Comment (osterman) [inline pkg/auth/identities/aws/user.go]: This hint is an explanation. An explanation is what happened. A hint is what to do about it.

File: pkg/auth/identities/aws/user.go [hunk 6]
```diff
@@ -223,76 +235,181 @@ func (i *userIdentity) writeAWSFiles(creds *types.AWSCredentials, region string)
...
+}
+
+// promptForNewCredentials prompts the user for new credentials and returns them.
+func (i *userIdentity) promptForNewCredentials(errorCode string) (*types.AWSCredentials, error) {
+	yamlMfaArn, _ := i.config.Credentials["mfa_arn"].(string)
+	newCreds, promptErr := PromptCredentialsFunc(i.name, yamlMfaArn)
+	if promptErr != nil {
+		log.Debug("Credential prompting failed", logKeyIdentity, i.name, "error", promptErr)
+		return nil, errUtils.Build(errUtils.ErrAwsCredentialsInvalid).
+			WithHint("Your AWS access keys have been rotated or revoked on the AWS side").
```
Comment (osterman) [inline pkg/auth/identities/aws/user.go]: This is the explanation.

File: pkg/auth/identities/aws/user.go [hunk 7]
```diff
@@ -223,76 +235,181 @@ func (i *userIdentity) writeAWSFiles(creds *types.AWSCredentials, region string)
...
+	}
+	log.Debug("New credentials provided via prompt", logKeyIdentity, i.name)
+	return newCreds, nil
+}
+
+// handleExpiredToken handles the case where the session token has expired.
+func (i *userIdentity) handleExpiredToken(apiErr smithy.APIError) (*types.AWSCredentials, error) {
+	log.Error("AWS session token expired", logKeyIdentity, i.name, "error_code", apiErr.ErrorCode())
+	return nil, errUtils.Build(errUtils.ErrAuthenticationFailed).
+		WithHint("AWS session token has expired").
```
Comment (osterman) [inline pkg/auth/identities/aws/user.go]: This is the explanation.

File: pkg/auth/identities/aws/user.go [hunk 8]
```diff
@@ -223,76 +235,181 @@ func (i *userIdentity) writeAWSFiles(creds *types.AWSCredentials, region string)
...
+		WithContext("identity", i.name).
+		WithContext("error_code", apiErr.ErrorCode()).
+		Err()
+}
+
+// handleAccessDenied handles the case where the user doesn't have permission to call GetSessionToken.
+func (i *userIdentity) handleAccessDenied(apiErr smithy.APIError) (*types.AWSCredentials, error) {
+	log.Error("Access denied when calling GetSessionToken", logKeyIdentity, i.name, "error_code", apiErr.ErrorCode())
+	return nil, errUtils.Build(errUtils.ErrAuthenticationFailed).
+		WithHint("Your IAM user does not have permission to call sts:GetSessionToken").
```
Comment (osterman) [inline pkg/auth/identities/aws/user.go]: This is the explanation.

**Comment by aknysh** [comment]
@coderabbitai full review please

**Comment by aknysh** [comment]
@coderabbitai full review please

**Comment by aknysh** [comment]
@coderabbitai full review please

**Comment by aknysh** [comment]
@coderabbitai full review please
